src/main/java/de/tum/in/www1/artemis/service/TextAssessmentQueueService.java

```diff
@@ -0,0 +1,158 @@
+package de.tum.in.www1.artemis.service;
+
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.*;
+import de.tum.in.www1.artemis.repository.TextClusterRepository;
+
+@Service
+@Profile("automaticText")
+public class TextAssessmentQueueService {
+
+    private final ParticipationService participationService;
+
+    private final TextClusterRepository textClusterRepository;
+
+    private Instant start;
+
+    private boolean running = false;
+
+    public TextAssessmentQueueService(ParticipationService participationService, TextClusterRepository textClusterRepository) {
+        this.participationService = participationService;
+        this.textClusterRepository = textClusterRepository;
+    }
+
+    /**
+     * Calculates the proposedTextSubmission for a given Text exercise
+     *
+     * @param textExercise the exercise for
+     * @throws IllegalArgumentException if textExercise isn't automatically assessable
+     * @return a TextSubmission with the highest information Gain if there is one
+     */
+    public Optional<TextSubmission> getProposedTextSubmission(TextExercise textExercise) {
+
+        if (!textExercise.isAutomaticAssessmentEnabled()) {
+            throw new IllegalArgumentException("The TextExercise is not   automatic assessable");
+        }
+
+        List<TextSubmission> textSubmissionList = participationService.findByExerciseIdWithEagerSubmittedSubmissionsWithoutResults(textExercise.getId()).parallelStream()
+                .map(StudentParticipation::findLatestTextSubmission).filter(Optional::isPresent).map(Optional::get).collect(Collectors.toList());
+        if (textSubmissionList.isEmpty()) {
+            return Optional.empty();
+        }
+        HashMap<TextBlock, Double> smallerClusterMap = calculateSmallerClusterPercentageBatch(textSubmissionList);
+        Optional<TextSubmission> best = textSubmissionList.stream().max(Comparator.comparingDouble(textSubmission -> calculateInformationGain(textSubmission, smallerClusterMap)));
+        return best;
+    }
+
+    /**
+     * Given an TextSubmission calculate the expected information gain
+     * based on the smallerClusterMap and addedDistancesMap
+     *
+     * @param textSubmission the textSubmission
+     * @param smallerClusterMap Map of TextBlocks to percentage of smaller clusters the TextBlock Cluster
+     * @return information gain for the TextSubmission
+     */
+    private double calculateInformationGain(TextSubmission textSubmission, HashMap<TextBlock, Double> smallerClusterMap) {
+        List<TextBlock> textBlocks = textSubmission.getBlocks();
+        double totalScore = 0.0;
+        for (TextBlock textBlock : textBlocks) {
+            if (textBlock.isAssessable()) {
+                continue;
+            }
+            if (textBlock.getCluster() == null) {
```

Comment:
same thing here, just be sure if this triggers queries.

Reply:
No queries get triggered here